viper/parser/parser.py

```diff
@@ -773,13 +789,51 @@ def check_list_type_match(provided):  # Check list types match.
 def pack_logging_data(expected_data, args, context):
     # Checks to see if there's any data
     if not args:
-        return ['seq'], 0, 0
+        return ['seq'], 0, None, 0
     holder = ['seq']
-    maxlen = len(args) * 32
+    maxlen = len(args) * 32  # total size of all packed args (upper limit)
+    dynamic_offset_counter = context.new_placeholder(BaseType(32))
+    dynamic_placeholder = context.new_placeholder(BaseType(32))
+
+    # Populate static placeholders.
+    placeholder_map = {}
     for i, (arg, data) in enumerate(zip(args, expected_data)):
         typ = data.typ
-        holder, maxlen = pack_args_by_32(holder, maxlen, arg, typ, context, context.new_placeholder(BaseType(32)))
-    return holder, maxlen, holder[1].to_list()[1][0]
+        placeholder = context.new_placeholder(BaseType(32))
+        placeholder_map[i] = placeholder
+        if not isinstance(typ, ByteArrayType):
+            holder, maxlen = pack_args_by_32(holder, maxlen, arg, typ, context, placeholder)
+
+    # Dynamic position starts right after the static args.
+    holder.append(LLLnode.from_list(['mstore', dynamic_offset_counter, maxlen]))
+
+    # Calculate maximum dynamic offset placeholders, used for gas estimation.
+    for i, (arg, data) in enumerate(zip(args, expected_data)):
```

Comment:
`for i, (arg, data) in enumerate(zip(args, expected_data)):`is the same as line 800, can we calculate the dynamic offset placeholders in the loop above?

Reply:
Because of the way maxlen possibly gets altered when we pack lists, the dynamic offset (maxlen right after the first loop) is required to start calculating the dynamic section, also if we put it one for loop I'd also be scared of perhaps not understanding what we are doing later on ?

Reply:
I don't see this as a compiler bottleneck and combining loops would make it less readable; let's leave it as it.  Thanks for the detailed explanation.